File: services/geoapi/geoapi/Storages/AgentStorage.vb
Imports System
Imports geoapi.Models
Imports geoapi.Utils
Imports MongoDB.Driver

Namespace Storages
    Class AgentStorage
        Implements IAgentStorage

        Private ReadOnly agents As IMongoCollection(Of AgentInfo)

        Public Sub New(ByVal settings As ISettings)
            Dim client = New MongoClient(settings.MongoDBConnectionString)
            Dim database = client.GetDatabase(settings.MongoDBName)
            agents = database.GetCollection(Of AgentInfo)(settings.AgentsCollectionName)
            agents.Indexes.DropAll()
            agents.Indexes.CreateOneAsync(Builders(Of AgentInfo).IndexKeys.Ascending(Function(__) __.AgentToken)).GetAwaiter().GetResult()
            agents.Indexes.CreateOne(Builders(Of AgentInfo).IndexKeys.Ascending("expireAt"), New CreateIndexOptions With {
                                        .ExpireAfter = TimeSpan.FromSeconds(30)
                                        })
        End Sub

        Public Sub IAgentStorage_AddAgent(agentInfo As AgentInfo) Implements IAgentStorage.AddAgent
            agents.InsertOne(agentInfo)
        End Sub

        Public Function IAgentStorage_GetAgent(agentId As String) As AgentInfo Implements IAgentStorage.GetAgent
            Return agents.Find(Function(agent) agent.AgentToken.Equals(agentId)).FirstOrDefault()
        End Function

    End Class
End Namespace

File: services/geoapi/geoapi/Storages/AgentStorage.cs
using System;
using geoapi.Models;
using geoapi.Utils;
using MongoDB.Driver;

namespace geoapi.Storages
{
    class AgentStorage : IAgentStorage
    {
        private readonly IMongoCollection<AgentInfo> agents;

        public AgentStorage(ISettings settings)
        {
            var client = new MongoClient(settings.MongoDBConnectionString);
            var database = client.GetDatabase(settings.MongoDBName);
            agents = database.GetCollection<AgentInfo>(settings.AgentsCollectionName);

            agents.Indexes.DropAll();

            agents.Indexes.CreateOneAsync(Builders<AgentInfo>.IndexKeys.Ascending(_ => _.AgentToken)).GetAwaiter().GetResult();
            agents.Indexes.CreateOne(Builders<AgentInfo>.IndexKeys.Ascending("expireAt"), new CreateIndexOptions { ExpireAfter = settings.TTL });
        }

        public AgentInfo GetAgent(string agentId)
        {
            return agents.Find(agent => agent.AgentToken == agentId).FirstOrDefault();
        }

        public void AddAgent(AgentInfo agentInfo)
        {
            agents.InsertOne(agentInfo);
        }
    }
}